### PR TITLE
[FLINK-13021][table][hive] unify catalog partition implementations

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

This PR unifies catalog partition implementations.

## Brief change log

- Unified `GenericCatalogPartition`, `HiveCatalogPartition`, and `AbstractCatalogPartition` into `CatalogPartitionImpl`
- moved string constant `1COMMENT` from `CatalogTableConfig` to `HiveCatalogConfig` as it's specific to the storage logic of `HiveCatalog`
- updated unit tests

## Verifying this change

This change is already covered by existing tests, such as `GenericInMemoryCatalogTest`, `HiveCatalogGenericMetadataTest`, `HiveCatalogHiveMetadataTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by bowenli86** [comment]
cc @xuefuz @lirui-apache @zjuwangg

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 1]
```diff
@@ -623,8 +622,10 @@ public void createPartition(ObjectPath tablePath, CatalogPartitionSpec partition
 		checkNotNull(partitionSpec, "CatalogPartitionSpec cannot be null");
 		checkNotNull(partition, "Partition cannot be null");
 
-		if (!(partition instanceof HiveCatalogPartition)) {
-			throw new CatalogException("Currently only supports HiveCatalogPartition");
+		boolean isGeneric = Boolean.valueOf(partition.getProperties().get(CatalogConfig.IS_GENERIC));
+
+		if (isGeneric) {
+			throw new CatalogException("Currently only supports non-generic CatalogPartition");
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: nit: The msg doesn't seem to be a complete sentence. Maybe something like "only xxx is supported currently". Same as below.

**Comment by bowenli86** [comment]
Good questions.

> 1. I saw a few configuration files for Hive (table/db, etc) and each has just few constants. I'm wondering if we should just have one file that covers all those configs.

Exactly. I've observed that too. I plan to merge them into a single `HiveCatalogConfig` in a following up.

> 2. When we create Catalog objects (table/db/partition), we remove "comment" from the property map. I'm wondering how that changes desc command output for those objects because ideally comment should be shown, at least for extended mode.

We only do that for table and partition.

- Hive's `Database` has a `description` field to store comment and we use that, thus that's the same output for 'DESC'
- Hive DDL for tables (e.g. `CREATE TABLE xx COMMENT xxx`) store comment as just a 'comment' property, so that's the same output for 'DESC'
- AFAIK, Hive DDL doesn't support partition comment, thus it shouldn't matter how we store it in properties

**Comment by xuefuz** [comment]
Okay. thanks for the explanation. I guess both don't matter much right now. Let's review on comments when we do DDL.

**Comment by bowenli86** [comment]
@xuefuz thanks for your review!

Due to the build queue is so congested now and my build request hasn't finished for almost a day, I built and tested locally and made sure it all passes.

Merging

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogPartition.java [hunk 2]
```diff
@@ -33,6 +33,13 @@
 	 */
 	Map<String, String> getProperties();
 
+	/**
+	 * Get comment of the partition.
+	 *
+	 * @return comment of the partition
+	 */
+	String getComment();
```
Comment (sunjincheng121) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogPartition.java]: Could you add the interface for Python API when you add the JAVA interface?

**Comment by sunjincheng121** [comment]
@bowenli86 please wait the CI passed when merging the PR!  

This commit caused the CI failed as the following error:

```
self = <pyflink.table.tests.test_catalog_completeness.CatalogPartitionAPICompletenessTests testMethod=test_completeness>

    def test_completeness(self):
>       self.check_methods()

pyflink/testing/test_case_utils.py:210: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pyflink.table.tests.test_catalog_completeness.CatalogPartitionAPICompletenessTests'>

    @classmethod
    def check_methods(cls):
        java_primary_methods = {'getClass', 'notifyAll', 'equals', 'hashCode', 'toString',
                                'notify', 'wait'}
        java_methods = PythonAPICompletenessTestCase.get_java_class_methods(cls.java_class())
        python_methods = cls.get_python_class_methods(cls.python_class())
        missing_methods = java_methods - python_methods - cls.excluded_methods() \
            - java_primary_methods
        if len(missing_methods) > 0:
            raise Exception('Methods: %s in Java class %s have not been added in Python class %s.'
>                           % (missing_methods, cls.java_class(), cls.python_class()))
E           Exception: Methods: set([u'getComment']) in Java class org.apache.flink.table.catalog.CatalogPartition have not been added in Python class <class 'pyflink.table.catalog.CatalogPartition'>.
```
I mentioned here because this is not the first time, you have merged the PR https://github.com/apache/flink/pull/8786  also caused the CI failed. 

http://mail-archives.apache.org/mod_mbox/flink-dev/201906.mbox/%3CCAJSjTKw4Qh6JJEPmTG6nUHQvdUwP1chLUMWvMES10R68AzUkXw@mail.gmail.com%3E

**Comment by dawidwys** [comment]
@bowenli86 A side note: you can also run the python tests locally. Build the whole project and invoke `./dev/lint-python.sh` in the `flink-python` module.

@sunjincheng121 Could we add a note somewhere how to run python tests locally? Maybe here: https://cwiki.apache.org/confluence/display/FLINK/Setting+up+a+Flink+development+environment ?

**Comment by sunjincheng121** [comment]
@dawidwys Thanks for the suggestion, setting up a Python Table API development environment is planned FLINK-12723, And I think we add how to run python test locally in https://cwiki.apache.org/confluence/display/FLINK/Setting+up+a+Flink+development+environment  is make more sense to me.

**Comment by bowenli86** [comment]
@sunjincheng121 @dawidwys sorry for it and thanks for the suggestion. Python test can't run locally on my machine currently, please see bug [FLINK-13085](https://issues.apache.org/jira/browse/FLINK-13085)

**Comment by sunjincheng121** [comment]
@bowenli86 Again:

What I'm emphasizing here is waiting for Travis to pass before you Merging PR. Whether you can test it locally or not, Flink master's Travis always working properly. Please don't let your merge commit affect other contributors'subsequent PRs.

As I mention above:  
 - The first time you merge commit (PR #8786) caused the contributor Haibo and zentol 's PR can not pass the CI and send the email in the mailing list. (Dian and I fix it)
- This time Wei Zhong' and Gary Gao's PR can not pass the CI test. ( Wei Zhong and I fix if)

In a short word:
I appreciate it if you can waiting for Travis change green before merging PR.

Best, Jincheng

**Comment by xuefuz** [comment]
Hi @sunjincheng121 ,

Thanks for sharing your thought. I agree with you in general on merging only after ci passes. However, we noticed recently that CI has been slow, queuing up a long list of PRs to be run against. With that, our PRs don't get a chance to run in days. That's why we took a shortcut occasionally. Nevertheless, we will follow the rule, but we have figured out a way to run ci on personal repos. Thus, we might still take shortcuts but will make sure CI passes before merging. Thanks for sharing your concern.

**Comment by dianfu** [comment]
@bowenli86 I just take a quick look at the issue reported in FLINK-13085, it's caused by the local Flink package is not consistent with the Python code as the interfaces of Java CSVTableSource has changed. So you should rebuild the Flink package before executing the Python tests.

**Comment by sunjincheng121** [comment]
Hi @xuefuz 

Our ultimate goal is to do our utmost to ensure that our merged code is as expected and does not affect the work of others. Waiting for CI to pass is a means, and if you have a better way to guarantee this, I agree! :)

If the way you say is verified by personal CI, then there is a reminder that tries to make the code of your own repo consistent with the target merge branch, such as the Flink master code, to avoid passing the local CI, but there is still a possibility of the problem when merging. :)

Thank you very much for agreeing on the rules of the merged code. I wish we have a happy contribution. 

Best,
Jincheng

**Pull request closed without merging**